File: go/onepiece/decider.go
package onepiece

type Decide[State any, Command any, Event any] func(state State, command Command) ([]Event, error)

type Evolve[State any, Event any] func(state State, event Event) State
type InitialState[State any] func() State

type IsTerminal[State any] func(state State) bool

type Decider[State any, Command any, Event any] struct {
	decide       Decide[State, Command, Event]
	evolve       Evolve[State, Event]
	initialState InitialState[State]
	isTerminal   IsTerminal[State]
}

func (d *Decider[State, Command, Event]) Decide(state State, command Command) ([]Event, error) {
	return d.decide(state, command)
}

func (d *Decider[State, Command, Event]) Evolve(state State, event Event) State {
	return d.evolve(state, event)
}

func (d *Decider[State, Command, Event]) InitialState() State {
	return d.initialState()
}

func (d *Decider[State, Command, Event]) IsTerminal(state State) bool {
	return d.isTerminal(state)
}

func NewDecider[State any, Command any, Event any](
	decide Decide[State, Command, Event],
	evolve Evolve[State, Event],
) *Decider[State, Command, Event] {
	decider := &Decider[State, Command, Event]{
		decide:       decide,
		evolve:       evolve,
		initialState: EmptyInitialState[State],
		isTerminal:   NeverTerminal[State],
	}

	return decider
}

func (d *Decider[State, Command, Event]) WithInitialState(initialState InitialState[State]) *Decider[State, Command, Event] {
	d.initialState = initialState
	return d
}

func (d *Decider[State, Command, Event]) WithIsTerminal(isTerminal IsTerminal[State]) *Decider[State, Command, Event] {
	d.isTerminal = isTerminal
	return d
}

func EmptyInitialState[State any]() State {
	var state State
	return state
}

func NeverTerminal[State any](_state State) bool {
	return false
}

File: unstable/rustlang/crates/infra/src/decider.rs
use std::result::Result;
use std::fmt::Debug;
use std::cmp::PartialEq;

pub type InitialState<State> = fn() -> State;
pub type IsTerminal<State> = fn(state: &State) -> bool;
pub type Decide<State, Command, Event, Error> = fn(state: &State, command: &Command) -> Result<Vec<Event>, Error>;
pub type Evolve<State, Event> = fn(state: &State, event: &Event) -> State;

pub struct Decider<State, Command, Event, Error> {
  initial_state: InitialState<State>,
  decide: Decide<State, Command, Event, Error>,
  evolve: Evolve<State, Event>,
  is_terminal: IsTerminal<State>,
}

impl<State, Command, Event, Error> Decider<State, Command, Event, Error>
  where
    Event: PartialEq + Debug,
    Error: PartialEq + Debug,
    State: PartialEq + Debug,
{
  pub fn new(
    decide: Decide<State, Command, Event, Error>,
    evolve: Evolve<State, Event>,
    initial_state: InitialState<State>,
    is_terminal: Option<IsTerminal<State>>,
  ) -> Self {
    Decider {
      decide,
      evolve,
      initial_state,
      is_terminal: is_terminal.unwrap_or(never_terminal),
    }
  }

  pub fn initial_state(&self) -> State {
    (self.initial_state)()
  }

  pub fn decide(&self, state: &State, command: &Command) -> Result<Vec<Event>, Error> {
    (self.decide)(state, command)
  }

  pub fn evolve(&self, state: &State, event: &Event) -> State {
    (self.evolve)(state, event)
  }

  pub fn is_terminal(&self, state: &State) -> bool {
    (self.is_terminal)(state)
  }
}

fn never_terminal<State>(_state: &State) -> bool {
  false
}
